Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -39,16 +39,71 @@ void EncodeObject(pb_ostream_t* stream,
 std::map<std::string, FieldValue> DecodeObject(pb_istream_t* stream);
 
 /**
- * Note that (despite the value parameter type) this works for bool, enum,
- * int32, int64, uint32 and uint64 proto field types.
- *
- * Note: This is not expected to be called direclty, but rather only via the
- * other Encode* methods (i.e. EncodeBool, EncodeLong, etc)
- *
- * @param value The value to encode, represented as a uint64_t.
+ * Docs TODO(rsgowman). But currently, this just wraps the underlying nanopb
+ * pb_ostream_t. Eventually, this might use static factory methods to create the
+ * underlying pb_ostream_t rather than directly passing it in.
  */
-void EncodeVarint(pb_ostream_t* stream, uint64_t value) {
-  bool status = pb_encode_varint(stream, value);
+class Writer {
+ public:
+  explicit Writer(pb_ostream_t* stream) : stream_(stream) {
+  }
+
+  /**
+   * Encodes a message type to the output stream.
+   *
+   * This essentially wraps calls to nanopb's pb_encode_tag() method.
+   *
+   * @param field_number is one of the field tags that nanopb generates based
+   * off of the proto messages. They're typically named in the format:
+   * <parentNameSpace>_<childNameSpace>_<message>_<field>_tag, e.g.
+   * google_firestore_v1beta1_Document_name_tag.
+   */
+  void EncodeTag(pb_wire_type_t wiretype, uint32_t field_number);
```

Comment:
I know this is _super_ annoying but I would expect that writers write, not encode.

Feel free to push this off with a todo or whatever until we've progressed down more of your review stack.

Reply:
Added TODO.